lib/AST/DiagnosticEngine.cpp

```diff
@@ -1129,7 +1129,14 @@ DiagnosticEngine::diagnosticStringFor(const DiagID id,
   auto defaultMessage = printDiagnosticNames
                             ? debugDiagnosticStrings[(unsigned)id]
                             : diagnosticStrings[(unsigned)id];
-  if (localization) {
+
+  // We load localization files only when the first diagnostic is requested, so
+  // there is no startup overhead for the compiler on projects without errors.
+  if (!getLocalizationLocale().empty()) {
+    if (!localization) {
+      setLocalization(getLocalizationLocale(), getLocalizationPath());
```

Comment:
Here is a problem with this approach - previously if `setLocalization` where to fail it would do so only once and set a pointer to `nullptr`, but now it would have to try and fail on every diagnostic produced by the program. I'd suggest an alternative approach (in the light of linker failures) - keep `setLocalization` the way it is and call it from `Frontend` but adjust `LocalizationProducer` to avoid loading data until first request. If it fails to load YAML, let's just fallback to using diagnostic strings from that point forward.


I'd suggest we rename `setLocalization` into `initializeLocalization` which would return a `std::unique_ptr<

Reply:
Do you mean we should go with the first approach I suggested which was 
- Make `YAMLLocalizationProducer::diagnostics` vector an optional.
- Remove the logic inside of the `YAMLLocalizationProducer` ’s constructor 
- In `getMessage` and `YAMLLocalizationProducer::forEachAvailable` we will check if diagnostics isn’t initialized, we will read the yaml file like we currently do in the constructor. (Move the logic of the YAMLLocalizationProducer ’s constructor to inside of this if statement)

I think since YamlProducer is only called in `setLocalization` that way we will fail once and will set `localization` pointer to nullptr. However, in other places in which we use YAMLLocalizationProducer like in tools and in unit tests... we will be trying to load the yaml file each time we try to get a message from maybe `YAMLLocalizationProducer::forEachAvailable`. So it's the same problem, as I understand... I think I will need some sort of flag so I don't try to initialize `diagnostics` again if it has failed.

I think we can overcome this with a simpler solution which is just add a boolean field to DiagnosticEngine that will check if we tried to call setLocalization or not, and in `diagnosticStringFor` we will check `if (!localization && !triedtoSetLocaliztion)`. I'm not sure however whether this approach considered "clean" or not and not sure also how I will be able to solve the linker failure without what I'm currently doing.

Reply:
> Do you mean we should go with the first approach I suggested which was

Note this is not only for YAML, `SerializedLocalizationProducer` shouldn't load anything either. I don't think it would be necessary to make `diagnostics` optional if there is a state recorded by `LocalizationProducer` regarding whether diagnostics have been initialized yet or not.

> I think we can overcome this with a simpler solution which is just add a boolean field to DiagnosticEngine that will check if we tried to call setLocalization or not, and in diagnosticStringFor we will check if (!localization && !triedtoSetLocaliztion). I'm not sure however whether this approach considered "clean" or not and not sure also how I will be able to solve the linker failure without what I'm currently doing.

How about we add such a flag to localization producer itself and provide a method for `DiagnosticEngine` to check whether it's valid or not?

Reply:
> How about we add such a flag to localization producer itself and provide a method for `DiagnosticEngine` to check whether it's valid or not?


This is a good idea. I think we can change `LocalizationProducer` and add a `bool validStatus` field. In the constructor we will make it equal `true` with `validStatus(true)`, and add a method to use in DiagEngine.
```
  /// Check if we tried to initialize LocalizationProducer or not.
  ///
  /// \returns true if we tried to initialize LocalizationProducer before, false otherwise.
  bool isValid() const { return validStatus; }
``` 
Finally, in `diagnosticStringFor` I can check with `if (!localization && !localization.get()->isValid())`.

Reply:
Yeah, you can define a boolean operator on `Localization` producer so that you can simplify check down to `if (auto producer =  localization.get())`